### PR TITLE
[MM-45118] Add insights endpoint for top dms

### PR DESCRIPTION
**Pull request by plant99**
#### Summary

Adds endpoint to get users' DM insights.

Sample request, response:
```
http://localhost:8065/api/v4/users/me/top/dms?page=0&per_page=10&time_range=28_day

{
	"has_next": false,
	"items": [
		{
			"post_count": 89,
			"second_participant": <json user object>,
			"outgoing_message_count": 23,
		},
		{
			"post_count": 14,
			"second_participant": <json user object>,
			"outgoing_message_count": 2,
		},
		{
			"post_count": 5,
			"second_participant":<json user object>,
			"outgoing_message_count": 5,
		}
	]
}
```

#### Ticket Link

https://mattermost.atlassian.net/browse/MM-45118
#### Release Note

```release-note
Added new API endpoints GET /api/v4/users/me/top/dms
```

### CODE REVIEW

File: api4/insights.go [hunk 1]
```diff
@@ -340,6 +329,36 @@ func getTopThreadsForUserSince(c *Context, w http.ResponseWriter, r *http.Reques
...
+func getTopDMsForUserSince(c *Context, w http.ResponseWriter, r *http.Request) {
+	user, err := c.App.GetUser(c.AppContext.Session().UserId)
+	if err != nil {
+		c.Err = err
+		return
+	}
+
+	startTime := model.StartOfDayForTimeRange(c.Params.TimeRange, user.GetTimezoneLocation())
+
+	topDMs, err := c.App.GetTopDMsForUserSince(c.AppContext.Session().UserId, &model.InsightsOpts{
```
Comment (mkraft) [inline api4/insights.go:342]:
```suggestion
	topDMs, err := c.App.GetTopDMsForUserSince(user.Id, &model.InsightsOpts{
```

File: api4/insights_test.go [hunk 2]
```diff
@@ -816,3 +816,87 @@ func TestGetTopThreadsForUserSince(t *testing.T) {
...
+
+	// get top dms for bu1
+	t.Run("get top dms for basic user 2", func(t *testing.T) {
+		th.LoginBasic2()
+		client = th.Client
+		topDMs, _, err := client.GetTopDMsForUserSince("today", 0, 100)
+		require.NoError(t, err)
+		require.Len(t, topDMs.Items, 2)
+		require.Equal(t, topDMs.Items[1].MessageCount, int64(1))
+	})
```
Comment (mkraft) [inline api4/insights_test.go:901]: Do you want to add a test that includes a bot account to ensure it's excluded from Top DMs? I can't think of any other edge cases... guest DMs should be listed as usual, they probably don't need an explicit test. What about deactivated users? I presume their DMs should continue to be listed for the time range, do you want to add a test to ensure they are?

File: api4/insights_test.go [hunk 3]
```diff
@@ -816,3 +816,87 @@ func TestGetTopThreadsForUserSince(t *testing.T) {
...
+	basicUser1 := th.BasicUser2
+
+	th.LoginBasic2()
+	client := th.Client
+	channelBu1, _, err := client.CreateDirectChannel(basicUser1.Id, basicUser1.Id)
+	require.NoError(t, err)
+
+	th.LoginBasic()
+	client = th.Client
+	channelBu, _, err := client.CreateDirectChannel(basicUser.Id, basicUser.Id)
```
Comment (mkraft) [inline api4/insights_test.go:840]: Is `channelBu` a direct channel with oneself? If so, should it be included in the results?

File: api4/insights_test.go (hunk 2 above)
Comment (plant99) [reply api4/insights_test.go:901]: Understood, will add the following two tests
1. bot DMs should be excluded.
2. Deactivated user's DM should be included

File: api4/insights_test.go (hunk 3 above)
Comment (plant99) [reply api4/insights_test.go:840]: Yes it is, I'll clean the variable names, I know it's confusing at the moment.

Comment (plant99) [reply api4/insights_test.go:840]: Oh, I think it should. But 1/5 and I'm open to correct it, because you, @laneycs , or @matthewbirtch  can help decide  on this better.

1. People talk less with theirselves, except for bookmarks, drafts and such.
2. Couldn't find a reason to exclude.

Comment (plant99) [reply api4/insights_test.go]: Awesome thank you both :) I'll exclude them and update the tests.

**Comment by BenCookie95** [comment]
@plant99 looks good but I will need the full user object in the response because I need to display their profile picture, preferred name and role

**Comment by BenCookie95** [comment]
@plant99 One other thing to note is that the modal also has a breakdown of messages sent vs messages received

**Comment by plant99** [comment]
Thanks for pointing the individual message contribution out @BenCookie95 , I keep referring to [this](https://github.com/mkraft/suite-users/blob/main/tech-specs/activity-and-insights-widgets/top-dms.md). Also noted on the `full user object`.

**Comment by BenCookie95** [comment]
> Thanks for pointing the individual message contribution out @BenCookie95 , I keep referring to [this](https://github.com/mkraft/suite-users/blob/main/tech-specs/activity-and-insights-widgets/top-dms.md). Also noted on the `full user object`.

That is outdated unfortunately, the figma has the latest stuff https://www.figma.com/file/fKyLulu6MUjQzQa9pBFFrC/Insights-and-Analytics?node-id=850%3A196684. Thanks for adding the user object!

File: api4/insights_test.go (hunk 3 above)
Comment (plant99) [reply api4/insights_test.go]: This is done.

**Comment by BenCookie95** [comment]
Looks great @plant99! One small thing I noticed is that we still count deleted messages in the totals

**Comment by plant99** [comment]
> Looks great @plant99! One small thing I noticed is that we still count deleted messages in the totals

Thanks for noticing ben, have fixed it.

**Comment by BenCookie95** [comment]
@plant99 It seems like the pagination isn't working properly, `hasNext` is always false even if there are items that should be on the next page

**Review by mkraft — APPROVED** [review]

**Review by BenCookie95 — APPROVED** [review]

**Review by furqanmlk — APPROVED** [review]
Working as expected

{
    "has_next": false,
    "items": [
        {
            "post_count": 5700,
            "outgoing_message_count": 2200,
            "second_participant": {
                "id": "36j9fx198jfa9mqr4cqb3iixuc",
                "last_picture_update": 1660854071361,
                "first_name": "First04e110e-asdfasdfasdfasdfasdfasdfasdfasdfasdfasdfasdfasdfasd",
                "last_name": "Last04e110e-asdfasdfasdfasdfasdfsdfsdfsdfsdfsdfsdfsdfsdfsdfsdfsd",
                "nickname": "Nickname04e110e",
                "username": "other04e110e-asdfasdfa",
                "position": "dasdfasdfas asdfasdfasdf asdfasdfasd asdfasdfasd asdfasdfasdf asdfasdfasdf asdfasdfasdf asdfasdfasdf asdfasdfasdf asdfasdfasdf a"
            }
        },
        {
            "post_count": 740,
            "outgoing_message_count": 540,
            "second_participant": {
                "id": "z518gazooir4pyhy9m6qnkff3e",
                "last_picture_update": 1660851449416,
                "first_name": "Firste7ee42f",
                "last_name": "Laste7ee42f",
                "nickname": "Nicknamee7ee42f",
                "username": "abc",
                "position": "asdfasdfasdfasd asdfasdfasdfasdf asdfasdfasdfasdf asdfasdfasdfa asdfasdfasdfasd asdfasdfasdfasdf asdfasdfasdfasdefasdgf sdafsdfa"
            }
        },
        {
            "post_count": 200,
            "outgoing_message_count": 200,
            "second_participant": {
                "id": "xbocxisifbdntqohk71kqwwy6r",
                "last_picture_update": 0,
                "first_name": "Firstf170900",
                "last_name": "Lastf170900",
                "nickname": "Nicknamef170900",
                "username": "otherf170900",
                "position": ""
            }
        },
        {
            "post_count": 100,
            "outgoing_message_count": 100,
            "second_participant": {
                "id": "f5399akiribb5fmtk1xuprn64c",
                "last_picture_update": 0,
                "first_name": "First1745515",
                "last_name": "Last1745515",
                "nickname": "Nickname1745515",
                "username": "admin1745515",
                "position": ""
            }
        },
        {
            "post_count": 50,
            "outgoing_message_count": 50,
            "second_participant": {
                "id": "zogajzfojbbp7rexcc91q3gf1h",
                "last_picture_update": 0,
                "first_name": "First68438fd",
                "last_name": "Last68438fd",
                "nickname": "Nickname68438fd",
                "username": "admin68438fd",
                "position": ""
            }
        },
        {
            "post_count": 50,
            "outgoing_message_count": 50,
            "second_participant": {
                "id": "gx16c1wi4bdztkg9413jg95hhc",
                "last_picture_update": 0,
                "first_name": "First34fc1e7",
                "last_name": "Last34fc1e7",
                "nickname": "Nickname34fc1e7",
                "username": "other34fc1e7",
                "position": ""
            }
        },
        {
            "post_count": 30,
            "outgoing_message_count": 30,
            "second_participant": {
                "id": "h34y1xraxt8rfmf9a1ant3xbme",
                "last_picture_update": 0,
                "first_name": "Firstfd15784",
                "last_name": "Lastfd15784",
                "nickname": "Nicknamefd15784",
                "username": "otherfd15784",
                "position": ""
            }
        },
        {
            "post_count": 10,
            "outgoing_message_count": 10,
            "second_participant": {
                "id": "1kdia5y673gwbdw1cknpkxpbbh",
                "last_picture_update": 0,
                "first_name": "Firsteff0e38",
                "last_name": "Lasteff0e38",
                "nickname": "Nicknameeff0e38",
                "username": "admineff0e38",
                "position": ""
            }
        }
    ]
}

**Comment by agnivade** [comment]
@plant99 - It doesn't seem like this endpoint was added to the load test tool. Is there any ticket scheduled to add that? Since this query is starting to show up as [slow queries](https://mattermost.atlassian.net/browse/MM-46601), this will start to affect performance and our load test will be unable to model that.

Please consider adding coverage to the load test tool as you keep adding new endpoints.

cc @agarciamontoro

**Comment by plant99** [comment]
Created a ticket https://mattermost.atlassian.net/browse/MM-46639, thanks for the reminder. I'll keep that in mind henceforth while adding new endpoints.
